aether-client-library/aether/client/oidc.py

```diff
@@ -0,0 +1,256 @@
+#
+# See the NOTICE file distributed with this work for additional information
+# regarding copyright ownership.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from bravado.requests_client import Authenticator, RequestsClient
+import requests
+
+from oauthlib.oauth2 import (
+    TokenExpiredError,
+    is_secure_transport,
+    InsecureTransportError
+)
+
+from requests_oauthlib import (
+    OAuth2Session,
+    TokenUpdated
+)
+import os
+import typing
+from urllib.parse import urlparse
+
+from .logger import LOG
+from .exceptions import AetherAPIException
+# don't force https for oauth requests
+os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = '1'
+
+
+class OauthClient(RequestsClient):
+    """Synchronous HTTP client implementation.
+    """
+
+    def __init__(self, authenticator, ssl_verify=True, ssl_cert=None):
+        """
+        :param ssl_verify: Set to False to disable SSL certificate validation.
+            Provide the path to a CA bundle if you need to use a custom one.
+        :param ssl_cert: Provide a client-side certificate to use. Either a
+            sequence of strings pointing to the certificate (1) and the private
+             key (2), or a string pointing to the combined certificate and key.
+        """
+        self.session = authenticator.session
+        self.authenticator = authenticator
+        self.ssl_verify = ssl_verify
+        self.ssl_cert = ssl_cert
+
+    def authenticated_request(self, request_params):
+        # type: (typing.Mapping[str, typing.Any]) -> requests.Request
+        return self.apply_authentication(
+            self.session.request(execute=False, **request_params))
+
+
+class OauthAuthenticator(Authenticator):
+
+    def __init__(
+        self,
+        server,
+        realm,
+        user=None,
+        pw=None,
+        keycloak_url=None,
+        offline_token=None
+    ):
+        self.realm = realm
+        self.host = urlparse(server).netloc
+        self.session = get_session(
+            server, realm, user, pw, keycloak_url, offline_token)
+
+    def bind_client(self, client):
+        client.session = self.session
+        client.authenticator = self
+
+    def get_spec(self, spec_url):
+        res = self.session.get(spec_url)
+        self.csrf = res.cookies.get('csrftoken')
+        LOG.debug(f'Set CSRFToken: {self.csrf}')
+        spec = res.json()
+        spec['host'] = f'{spec["host"]}/{self.realm}/kernel'
```

Comment:
I would never hard-code `kernel`... we are using the `GATEWAY_ID` )or something like that env var.

Reply:
I stuck it into the `Client` constructor as an optional kwarg.